art/defences/preprocessor/inverse_gan.py

```diff
@@ -0,0 +1,215 @@
+            z_i_reshaped = np.reshape(z_i, [batch_size, self.gan.encoding_length])
+            loss = self.loss(z_i_reshaped, x)
+
+            return loss
+
+        options = {}
+
+        options_allowed_keys = [
+            "disp",
+            "maxcor",
+            "ftol",
+            "gtol",
+            "eps",
+            "maxfun",
+            "maxiter",
+            "iprint",
+            "callback",
+            "maxls",
+        ]
+
+        for key in kwargs:
+            if key not in options_allowed_keys:
+                raise KeyError(
+                    "The argument `{}` in kwargs is not allowed as option for `scipy.optimize.minimize` using "
+                    '`method="L-BFGS-B".`'.format(key)
+                )
+
+        options.update(kwargs)
+
+        optimized_z_encoding_flat = minimize(func_loss, initial_z_encoding, jac=func_gen_gradients, method="L-BFGS-B",
+                                             options=options)
+        optimized_z_encoding = np.reshape(optimized_z_encoding_flat.x, [batch_size, self.gan.encoding_length])
+
+        y = self.gan.predict(optimized_z_encoding)
+
+        return y
+
+    def loss(self, z, image_adv):
+        """
+        Given a encoding z, computes the loss between the projected sample and the original sample
+        :param z: encoding z
+        :type z: `np.ndarray`
+        :param image_adv:
+        :type image_adv: `np.ndarray`
+        :return: The loss value
+        """
+
+        logging.info("Calculating Loss")
+
+        loss = self._sess.run(self._loss,
+                              feed_dict={self.gan.input_ph: z, self._image_adv: image_adv})
+
+        return loss
+
+    def loss_gradient(self, z_encoding, y):
+        """
+        Compute the gradient of the loss function w.r.t. a `z_encoding` input within a GAN against a
+        corresponding adversarial sample
+        :param z_encoding:
+        :type z_encoding: `np.ndarray`
+        :param y: Target values of shape (nb_samples, nb_classes)
+        :type y: `np.ndarray`
+        :return: Array of gradients of the same shape as `z_encoding`.
+        :rtype: `np.ndarray`
+        """
+
+        logging.info("Calculating Gradients")
+
+        gradient = self._sess.run(self._grad,
+                                  feed_dict={self._image_adv: y,
+                                             self.gan.input_ph: z_encoding})
+
+        return gradient
+
+    @property
+    def apply_fit(self):
+        """
+        do nothing.
+        """
+        pass
+
+    @property
+    def apply_predict(self):
+        """
+        do nothing.
+        """
+        pass
+
+    def estimate_gradient(self, x, grad):
+        """
+        do nothing.
+        """
+        pass
+        # return grad
+
+    def fit(self, x, y=None, **kwargs):
+        """
+        No parameters to learn for this method; do nothing.
+        """
+        pass
+
+
+class DefenseGan(InverseGanDefense):
```

Comment:
I think we should move `DefenseGan` to a separate file `defense_gan.py`

Reply:
ok sure